su1/errors-handling.go

```diff
@@ -0,0 +1,44 @@
+package su1
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+)
+
+func ErrorsHandling() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	tasks := make(chan Task)
+
+	done, errors := WorkerPool(ctx, 3, tasks)
+
+	for i := 0; i < 10; i++ {
+		idx := i
+
+		tasks <- func() error {
+			<-time.After(time.Second * 1)
+
+			if idx%2 == 0 {
+				return fmt.Errorf("an error occurred within a task %d\n", idx)
+			}
+
+			println("Task " + strconv.Itoa(idx) + " is completed")
+
+			return nil
+		}
+	}
+
+	close(tasks)
+
+	for err := range errors {
```

Comment:
Here we are stopping the execution on the first error. We can and should use `select` instruction.

```
select {
case err := <- errors:
	if err != nil {
		log.Println(err.Error())
		cancel()
	}
case <-done:
	log.Println("completed without errors")
}
```

Why can't we listen for `errors` first and then for the `done` channel? Because the `done` channel is your main indicator that the program is finished and it has the same priority as the `errors` channel, hence we should listen to them simultaneously.